libc-bottom-half/headers/private/wasi/descriptor_table.h

```diff
@@ -106,18 +106,53 @@ typedef struct {
 	udp_socket_state_t state;
 } udp_socket_t;
 
+typedef struct {
+// Stream for contents of directory
+        filesystem_own_directory_entry_stream_t directory_stream;
+// File handle for the directory being listed
+// (so that metadata hashes can be accessed).
+        filesystem_borrow_descriptor_t directory_file_handle;
+} directory_stream_entry_t;
+
+// Stream representing an open file.
+typedef struct {
+// File was opened for reading
+         bool readable;
+// File was opened for writing
+         bool writable;
+         filesystem_borrow_descriptor_t file_handle;
+} file_t;
+
+typedef struct {
+        streams_borrow_input_stream_t read_stream;
+        streams_borrow_output_stream_t write_stream;
+        // Current position in stream, relative to the beginning of the *file*, measured in bytes
+        off_t offset;
+        // When the stream is closed, the caller should
+        // replace this entry in the table with the file handle
+        file_t file_info;
+} file_stream_t;
+
 // This is a tagged union. When adding/removing/renaming cases, be sure to keep the tag and union definitions in sync.
 typedef struct {
 	enum {
 		DESCRIPTOR_TABLE_ENTRY_TCP_SOCKET,
 		DESCRIPTOR_TABLE_ENTRY_UDP_SOCKET,
+                DESCRIPTOR_TABLE_ENTRY_FILE_HANDLE,
+                DESCRIPTOR_TABLE_ENTRY_DIRECTORY_STREAM,
+                DESCRIPTOR_TABLE_ENTRY_FILE_STREAM
```

Comment:
Mix of tabs/spaces

Reply:
Replaced tabs with spaces in [5932e2f](https://github.com/WebAssembly/wasi-libc/pull/606/commits/5932e2fc4d0a9d5443af938e89331a09f429f2c8)